hazelcast-client-new/src/main/java/com/hazelcast/client/txn/proxy/xa/XATransactionProxy.java

```diff
@@ -0,0 +1,151 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.client.txn.proxy.xa;
+
+import com.hazelcast.client.connection.nio.ClientConnection;
+import com.hazelcast.client.impl.HazelcastClientInstanceImpl;
+import com.hazelcast.client.impl.protocol.ClientMessage;
+import com.hazelcast.client.impl.protocol.parameters.TransactionCreateResultParameters;
+import com.hazelcast.client.impl.protocol.parameters.XATransactionCommitParameters;
+import com.hazelcast.client.impl.protocol.parameters.XATransactionCreateParameters;
+import com.hazelcast.client.impl.protocol.parameters.XATransactionPrepareParameters;
+import com.hazelcast.client.impl.protocol.parameters.XATransactionRollbackParameters;
+import com.hazelcast.client.spi.impl.ClientInvocation;
+import com.hazelcast.transaction.TransactionException;
+import com.hazelcast.transaction.TransactionNotActiveException;
+import com.hazelcast.transaction.impl.Transaction;
+import com.hazelcast.transaction.impl.xa.SerializableXID;
+import com.hazelcast.util.Clock;
+import com.hazelcast.util.EmptyStatement;
+import com.hazelcast.util.ExceptionUtil;
+
+import javax.transaction.xa.Xid;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+
+import static com.hazelcast.transaction.impl.Transaction.State.ACTIVE;
+import static com.hazelcast.transaction.impl.Transaction.State.COMMITTED;
+import static com.hazelcast.transaction.impl.Transaction.State.NO_TXN;
+import static com.hazelcast.transaction.impl.Transaction.State.PREPARED;
+import static com.hazelcast.transaction.impl.Transaction.State.ROLLED_BACK;
+import static com.hazelcast.transaction.impl.Transaction.State.ROLLING_BACK;
+
+public class XATransactionProxy {
+
+    private final HazelcastClientInstanceImpl client;
+    private final ClientConnection connection;
+    private final SerializableXID xid;
+    private final int timeout;
+
+    private Transaction.State state = NO_TXN;
+    private volatile String txnId;
+    private long startTime;
+
+    public XATransactionProxy(HazelcastClientInstanceImpl client, ClientConnection connection, Xid xid, int timeout) {
+        this.client = client;
+        this.connection = connection;
+        this.timeout = timeout;
+        this.xid = new SerializableXID(xid.getFormatId(), xid.getGlobalTransactionId(), xid.getBranchQualifier());
+    }
+
+    void begin() {
+        try {
+            if (state == ACTIVE) {
```

Comment:
no, can be commit/rollbacked from a different thread, same for `XATransactionImpl`.
They both needs concurrency controls.